vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClient.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2021 Red Hat, Inc.
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *  The Eclipse Public License is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  The Apache License v2.0 is available at
+ *  http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.ext.web.client;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpClient;
+import io.vertx.ext.web.client.spi.CacheStore;
+import io.vertx.ext.web.client.impl.WebClientBase;
+import io.vertx.ext.web.client.impl.CachingWebClientImpl;
+import io.vertx.ext.web.client.impl.cache.CacheManager;
+
+/**
+ * An asynchronous cache aware HTTP / HTTP/2 client called {@code WebClientCache}.
+ * <p>
+ * This client wraps a {@link WebClient} and makes it cache aware adding features to it:
+ * <ul>
+ *   <li>Cache-Control header parsing</li>
+ *   <li>Response serialization and deserialization for the {@link CacheStore}</li>
+ *   <li>Freshness checking</li>
+ * </ul>
+ * <p>
+ * The client honors the cache headers:
+ * <ul>
+ *  <li>Cache-Control with the following properties understood:
+ *    <ul>
+ *      <li>public</li>
+ *      <li>private</li>
+ *      <li>no-cache</li>
+ *      <li>no-store</li>
+ *      <li>max-age</li>
+ *      <li>s-maxage</li>
+ *    </ul>
+ *  </li>
+ *  <li>Expires</li>
+ *  <li>ETag</li>
+ *  <li>Vary</li>
+ * </ul>
+ * <p/>
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public interface CachingWebClient extends WebClient {
```

Comment:
I don't think there is a real interest to have a specific `CachingWebClient` interface that extends `WebClient` as it does not have specific operations.

Reply:
Does this mean it should not extend `WebClient`, and if so what should it do instead, or are you suggesting that we add the cache features into the core WebClient itself and just make it configurable?

Reply:
yes the interface should not extend the WebClient but the implementation should still extend WebClient of course. It is just that we do not need a new functionnal interface for this.

Reply:
so the implementation that wraps the actual WebClient still implements WebClient

Reply:
Ok. So how would someone get the caching implementation? Should it be via config, or a few some new functions on the `WebClient` interface?